sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -186,6 +187,69 @@ trait Partitioning {
   }
 }
 
+object Partitioning {
+  /**
+   * Gets as input the `Partitioning` of an expressions and returns the valid `Partitioning`s for
+   * the node w.r.t its output and its expressions.
+   */
+  def updatePartitioningWithNewOutput(
+      inputPartitioning: Partitioning,
+      expressions: Seq[NamedExpression],
+      outputSet: AttributeSet): Partitioning = {
+    inputPartitioning match {
+      case partitioning: Expression =>
+        val exprToEquiv = partitioning.references.map { attr =>
+          attr -> expressions.filter(e =>
+            CleanupAliases.trimAliases(e).semanticEquals(attr))
+        }.filterNot { case (attr, exprs) =>
+          exprs.size == 1 && exprs.forall(_ == attr)
+        }
+        val initValue = partitioning match {
+          case PartitioningCollection(partitionings) => partitionings
+          case other => Seq(other)
+        }
+        val validPartitionings = exprToEquiv.foldLeft(initValue) {
+          case (partitionings, (toReplace, equivalents)) =>
+            if (equivalents.isEmpty) {
+              partitionings.map {
+                case hp: HashPartitioning if hp.references.contains(toReplace) =>
+                  UnknownPartitioning(hp.numPartitions)
+                case rp: RangePartitioning if rp.references.contains(toReplace) =>
```

Comment:
since we need to handle different partitioning differently, shall we add a method to `Partitioning`?